File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null){
            return null;
        }
        if(root == p || root == q){
            return root;
        }

        TreeNode pleft = lowestCommonAncestor(root.left,p,q);
        TreeNode pright = lowestCommonAncestor(root.right,p,q);

        if(pleft != null && pright != null){
            return root;
        }

        return pleft == null ? pright : pleft ;


        
    }
}

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
                
            def helper(root):    
                if not root:
                   return None
                
                if root == p or root == q:
                    return root
                
                left = helper(root.left)
                right = helper(root.right)

                if left and right:
                    return root
                
                if not left:
                    return right
                return left
            
            return helper(root)
                 
                
                
        